### ISSUE TITLE
Action cache is not working in some cases

### ISSUE THREAD

**Issue by ulfjack** [bazelbuild/bazel#2490]
After I build a new Bazel binary, I get almost no action cache hits, even if the binaries are identical (I think). This happens on MacOS at least. I'm investigating.

**Comment by ulfjack** [comment]
It happens even after a shutdown, for things as simple as bazel build //third_party/ijar.

**Comment by ulfjack** [comment]
It looks like the round-trip through the on-disk representation is buggy. I get the exact same digest on subsequent runs, but it's always not matching the digest read from the on-disk action cache.

**Comment by ulfjack** [comment]
Previous analysis was incorrect - it seems to only apply to some C++ compile actions, and seems likely to be related to input discovery / subsetting.

**Comment by ulfjack** [comment]
Here's the action cache entry we write:

    12e745723ce3d938d4f9c946e8754342 97d3f6702bf302c459e19b25c26bfb9e
      actionKey = a9f1d0ed60bfa5141202b858a610cd25
      digestKey = 97d3f6702bf302c459e19b25c26bfb9e
      bazel-out/local-fastbuild/bin/foo/_objs/foo/foo/foo.pic.d
      bazel-out/local-fastbuild/bin/foo/_objs/foo/foo/foo.pic.o
      bazel-out/local-fastbuild/internal/_middlemen/_S_Sfoo_Cfoo-null
      external/local_config_cc/cc_wrapper.sh
      foo/foo.cc
      foo/foo.h

Here's the (partial) entry we compare it to:

    <snip>
      bazel-out/local-fastbuild/bin/foo/_objs/foo/foo/foo.pic.o
      bazel-out/local-fastbuild/bin/foo/_objs/foo/foo/foo.pic.d
      external/local_config_cc/cc_wrapper.sh
      foo/foo.cc
      bazel-out/local-fastbuild/internal/_middlemen/_S_Sfoo_Cfoo-null

Order of files does not matter, but note that we're missing foo/foo.h.

**Comment by ulfjack** [comment]
It seems like this is causing a lot of recompiling for C++ applications after a shutdown.

**Comment by ulfjack** [comment]
This is draining my battery because every experiment I run rebuilds a lot of C++ code. However, I think ActionExecutionFunction.collectInputs is misbehaving.

**Comment by ulfjack** [comment]
Well, that diagnosis was incorrect as well. New theory: in Bazel, CppCompileAction returns true for inputsKnown(), because include scanning is disabled (and not implemented). However, it still updates the action inputs after execution. At the same time ActionExecutionFunction is only using the files resolved from the cache for the action cache lookup if inputsKnown() returns false.

**Comment by ulfjack** [comment]
I still think my last explanation is correct, but it doesn't seem to be sufficient - even with the inputsKnown() method fixed, I still see unnecessary action execution. My battery is now down to <30%. :-(

**Comment by ulfjack** [comment]
So, looks better, but I get a null md5 digest for external/local_config_cc/cc_wrapper.sh.

**Comment by ulfjack** [comment]
We have two artifacts for cc_wrapper.sh, and one of them doesn't actually exists, and so does not have a md5:
Artifact:[/Users/ulfjack/Google/bazel[source]]external/local_config_cc/cc_wrapper.sh=singleton marker artifact value (1638529830)

 Artifact:[/private/var/tmp/_bazel_ulfjack/ea721f29c7ff430b1d606476a9c89991[source]]external/local_config_cc/cc_wrapper.sh=RegularFileArtifactValue{digest=[44, 94, -17, 119, -89, -55, 22, -35, -31, -123, -125, 122, 4, -67, 63, 25], size=2730}

What I don't understand is that we somehow seem end up with the first artifact before running the action and with the second artifact after. In any case, if it only affects cc_wrapper.sh, then it may work on Linux, where we don't use a wrapper.

**Comment by ulfjack** [comment]
Apparently, the code that's resolving artifacts from exec paths is ending up with the wrong artifact, while the action has the correct artifact as input, and we use the resolved artifacts to find an action cache hit, where we ignore that one of the inputs doesn't actually exist.

**Comment by ulfjack** [comment]
It looks like ArtifactFactory cannot correctly handle artifacts under external/.

**Comment by ulfjack** [comment]
Ok, I found a workaround, which is to treat files under external/ as output files in ActionCacheChecker.getCachedInputs:

    @@ -319,6 +340,11 @@ public class ActionCacheChecker {
             allowedDerivedInputsMap.put(derivedInput.getExecPath(), derivedInput);
           }
         }
    +    for (Artifact input : action.getMandatoryInputs()) {
    +      if (input.getExecPath().startsWith(new PathFragment("external"))) {
    +        allowedDerivedInputsMap.put(input.getExecPath(), input);
    +      }
    +    }
     
         List<Artifact> inputArtifacts = new ArrayList<>();
         List<PathFragment> unresolvedPaths = new ArrayList<>();

**Comment by ulfjack** [comment]
The other part of the fix is this:

    +++ b/src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileAction.java
    @@ -334,7 +334,7 @@ public class CppCompileAction extends AbstractAction
         this.shouldPruneModules = shouldPruneModules;
         this.usePic = usePic;
         this.useHeaderModules = useHeaderModules;
    -    this.inputsKnown = !shouldScanIncludes;
    +    this.inputsKnown = !shouldScanIncludes && !shouldPruneModules && !cppSemantics.needsDotdInputPruning();
         this.cppCompileCommandLine =
             new CppCompileCommandLine(
                 sourceFile, dotdFile, copts, coptsFilter, features, variables, actionName);

**Comment by ulfjack** [comment]
For completeness: for actions whose inputs can change over their lifetime, the action cache behaves differently than for actions who don't. The action cache uses inputsKnown() to determine whether the inputs can change or not.

If the inputs can change, the action cache reads the list of input files from the action cache and uses them to check if the action can be cached. However, the action cache only stores relative paths, so the action cache needs to resolve them to artifacts.

This is tricky - there are three kinds of artifacts. For output files, it considers the list of declared inputs of the current action to find the corresponding artifacts. For source files, it needs to find the correct --package_path entry and package that's responsible for that source file. That uses Skyframe to look up the corresponding locations. This generally all works.

Unfortunately, the code treats files in external/ as if they were source files, even though they are actually generated files (more or less - at least they are not in the source tree). Thus, it ends up with the wrong artifact when it checks the action cache before execution - the corresponding source file does not exist, and so the action cache check silently fails. Silent failure is necessary because a user might have legitimately deleted a file.

The code snippet above makes the action cache checker treat files in external/ as if they were generated files, i.e., it requires that they are declared inputs to the current action, and uses the relative path to look them up.

**Comment by ulfjack** [comment]
I wonder if 840acb10fb91b04e87d45f5627f888464899e577 is related.

**Comment by kchodorow** [comment]
The files in <output_base>/external/ should be treated as source files, I think.

**Comment by ulfjack** [comment]
For the purpose of the C++ compile action, the external/ paths are special, because they are not in the source tree, and they are also not under any of the output roots, so - right now - neither code path is correct. However, it so happens that I can put the external/ paths into the same code path as generated files and that works. Also note that `external/local_config_cc/cc_wrapper.sh` in particular happens to be a generated file, even if generated by the workspace.

**Comment by ulfjack** [comment]
Isn't it?

**Comment by jmmv** [comment]
Hah! I have spent three days diagnosing this same issue and, now that I had found a cause, knew what to look for -- which led me to this bug. Let me paste some of the details from my investigation, which include a much simpler test case to iterate on this.

Consider these files:

    -- foo/BUILD --
    cc_library(name = "foo", srcs = ["foo.cc", "foo.h"])

    -- foo/foo.cc --
    #include "foo/foo.h"
    void foo(void) { }

    -- foo/foo.h --
    // Nothing.

Then, the following operations:

    bazel build --explain log --verbose_explanations --check_visibility //foo:foo
    bazel build --explain log --verbose_explanations --check_visibility //foo:foo
    bazel build --explain log --verbose_explanations --nocheck_visibility //foo:foo

The first command should build things as usual. The second command will yield an empty log file (disregarding the details for `stable-status.txt`). The third command, however, will report that `foo.cc` has changed and needs to be recompiled.

Removing the `#include "foo/foo.h"` from `foo.cc` makes the problem go away.

Focusing on `ActionCacheChecker`, in `validateArtifacts` during the third command above, the `Entry` object properly tracks `foo.h` as a path... but the `actionInputs` *do not* contain `foo.h`. Thus the newly computed digest does not match what we recorded in the action cache and miss, triggering the rebuild. Tricking the code into consider the header file causes the hashes to match and a cache hit.

**Comment by janakdr** [comment]
@jmmv : @ulfjack's issue and yours seem to be orthogonal. I don't think his fix will fix your issue (although conceivably a fix to your issue might fix his).

**Comment by ulfjack** [comment]
Hmm? The --nocheck_visibility flag triggers re-analysis, which in turn triggers an action cache check for the new actions even if the actions are identical. Fixing the action cache check for C++ actions seems like it should improve things significantly (compared to actually running the compiler).

**Comment by jmmv** [comment]
I'm happy to move the issue I found to a separate bug if there is more than one fix to be made (and it seems like that's the case). The reason I added the details here is because @ulfjack troubleshooting steps looked very similar to mine and I wondered if they were obscured by what I found.

**Comment by janakdr** [comment]
I agree that both issues are related to the action cache and Bazel. I'm just saying that @ulfjack's issue appears to be specific to external files, and @jmmv's is about all C++ compilation. Perhaps the latter will resolve the former (although I doubt it), but not vice versa.

**Comment by ulfjack** [comment]
I actually found two bugs, one in CppCompileAction, and another related to our handling of //external. I had to fix both to get it to work on my Laptop.

**Comment by ulfjack** [comment]
So, assuming that Kristina is right and that these files should be treated as source files, I dug even deeper. The symptom is that ArtifactFactory doesn't return the correct Artifact. What's likely happening is that the PackageRootResolver is returning the wrong package root. PackageRootResolver is implemented by SkyframeExecutor.getArtifactRoots, which contains this code:

    PathFragment parent = Preconditions.checkNotNull(
        execPath.getParentDirectory(), "Must pass in files, not root directory");
    packageKeys.add(ContainingPackageLookupValue.key(
        PackageIdentifier. *createInMainRepo* (parent)));

That's incorrect. It tries to look for the corresponding path in the main repo, even though it's actual in an external repo. Maybe we should check for an "external" prefix and then create the right repo identifier instead:

    if (parent.startsWith(new PathFragment(Label.EXTERNAL_PATH_PREFIX)) && parent.segmentCount() >= 2) {
      RepositoryName repository = RepositoryName.create("@" + parent.getSegment(1));
      packageKeys.add(ContainingPackageLookupValue.key(
          PackageIdentifier.create(repository, parent.subFragment(2, parent.segmentCount()))));
    } else {
      packageKeys.add(ContainingPackageLookupValue.key(
          PackageIdentifier.createInMainRepo(parent)));
    }

**Comment by ulfjack** [comment]
I accidentally tried to reproduce this with an old bazel binary - pre 4ec02263fa630e97ce9a749a02efe9cf64838690 - and couldn't reproduce, so this problem must have been introduced in the last ~25 days.

**Comment by ulfjack** [comment]
It fails at 54223b019817c329d7bff1c174d497ef3d64695a, at least, so somewhere between 16 and 25 days ago.

**Comment by ulfjack** [comment]
No, my bad. It works at 54223b0.

**Comment by ulfjack** [comment]
It fails at 840acb1, so that's where this bug got introduced. @lberki

**Comment by ittaiz** [comment]
@Ulf- binary search? ;)

**Comment by ulfjack** [comment]
Meh. We have two Skyframe implementations of PackageRootResolver, and I was looking at the wrong one. The correct one is ActionExecutionFunction.PackageRootResolverWithEnvironment.

**Comment by ulfjack** [comment]
Well, this isn't working. I get the wrong source root back from the root resolution, except I don't know why. Here's the source root I get: /private/var/tmp/_bazel_ulfjack/fe875d5ca7c5e5732f6aae397d7cdf99/external/local_config_cc[source]

Here's the source root of the artifact that's written to the cache:
/private/var/tmp/_bazel_ulfjack/fe875d5ca7c5e5732f6aae397d7cdf99[source]

**Comment by ulfjack** [comment]
Ok, I have a fix, but I'm not sure this is correct. :-/

**Comment by kchodorow** [comment]
Please cc me on the fix, /private/var/tmp/_bazel_ulfjack/fe875d5ca7c5e5732f6aae397d7cdf99/external/local_config_cc will be the correct source root once the execroot change is in (just waiting on TF to merge https://github.com/tensorflow/tensorflow/pull/7165!).

**Comment by ulfjack** [comment]
I'd be more than happy for you to fix this, @kchodorow. :-P

**Comment by ulfjack** [comment]
@kchodorow how do you plan to switch the source root behavior? Is there a flag or something?

**Comment by ulfjack** [comment]
Is the plan to roll forward 82d43279f93d95e4c41b4bc598a3cc05ddd1ae1a? Unfortunately, I can't easily patch it in right now as it doesn't apply cleanly anymore.

**Comment by kchodorow** [comment]
Basically yes. The internal # is 143777249, if you want to check it out. 

(Now just waiting on https://github.com/tensorflow/serving/pull/321...)

**Comment by kchodorow** [comment]
...and I accidentally reverted the CL ($*%&$) so now it's 147374155.

**Comment by werkt** [comment]
We got around this issue of the on-disk-roundtrip-exclude-invalidation by adding a symlink to the external execroot in the workspace root.  We can post a diff if that's an acceptable workaround (my guess is not).

**Comment by ulfjack** [comment]
I'm still on it, even if it takes a bit longer. It looks like there's a third place in the code that's doing artifact discovery and that's also getting incorrect results. :-/

**Comment by ulfjack** [comment]
And that one got rolled back as well. I have a complete fix, but the change I committed depends on Kristina's change, which got rolled back. I think I can make it so it works both with and without Kristina's change... next week.

**Comment by werkt** [comment]
With my workaround, and changing inputsKnown to not be dependent upon shouldScanIncludes (the negation thereof) and just always false for CppCompileAction, the original patch works fine for me.

**Comment by damienmg** [comment]
@ulfjack: will you have time to work on the fix, should we ask @mhlopko to take over?

**Comment by ulfjack** [comment]
I've sent a fix for review to Kristina. Hopefully it'll be merged later today.

**Comment by werkt** [comment]
I was just looking at the test from the prune fix (`src/test/shell/integration/cpp_test.sh`), and realized that the test is relying on `Compiling a/a.cc` appearing in the output, which, unless I'm mistaken, is a timing artifact, and may not appear when the compile actually takes place  - it should probably be using `-s` and looking for something indicating the compile activity

**Comment by ulfjack** [comment]
I thought I had a fix, but I was wrong. I've spent a couple more hours tracing the code and I think I've got it now. I'm running all of our tests again.

**Comment by werkt** [comment]
I have put together a test case for some unnecessary recompilations that are taking place that fails under the #2372 patch, but succeeds when inputsKnown is decoupled from shouldScanIncludes in CppCompileAction, see this [diff](https://gist.github.com/werkt/cf53b5201a8b0be19e462b3566f1cdfd)

**Comment by mhlopko** [comment]
cc @lberki

**Comment by werkt** [comment]
After testing further with a daemon shutdown between builds, it was only the inputsKnown initially false for compiles that kept recompilations from occurring after a build, for any cc_test *or* cc_binary compilation action.

**Comment by ulfjack** [comment]
We also need to add a regression test, maybe based on @werkt's patch.

**Comment by werkt** [comment]
Regression test based on shutdown to incur on disk roundtrip.

https://github.com/bazelbuild/bazel/pull/2586

I'm investigating the action reevalulation prevention that is occurring in the odd corner case with cc_test compilations + crosstool + interleaved query.  Will post another PR when I get there.

**Comment by jmmv** [comment]
Is this fully fixed? I tried the repro in https://github.com/bazelbuild/bazel/issues/2490#issuecomment-278660955 with a head-built Bazel and still hit the problem. Should I file a separate bug for that issue?
